nav2_recoveries/src/spin.cpp

```diff
@@ -49,50 +49,50 @@ Spin::~Spin()
 
 Status Spin::onRun(const std::shared_ptr<const SpinAction::Goal> command)
 {
-  double yaw, pitch, roll;
-  tf2::getEulerYPR(command->target.quaternion, yaw, pitch, roll);
+  double pitch, roll;
+  tf2::getEulerYPR(command->target.quaternion, command_yaw_, pitch, roll);
 
   if (roll != 0.0 || pitch != 0.0) {
-    RCLCPP_INFO(node_->get_logger(), "Spinning on Y and X not supported, "
-      "will only spin in Z.");
+    RCLCPP_INFO(node_->get_logger(), "Spinning around Y and X axes are not supported, "
+      "will only spin around Z axis.");
   }
 
-  RCLCPP_INFO(node_->get_logger(), "Currently only supported spinning by a fixed amount");
+  if (!getRobotPose(initial_pose_)) {
+    RCLCPP_ERROR(node_->get_logger(), "initial robot pose is not available.");
+    return Status::FAILED;
+  }
 
-  start_time_ = std::chrono::system_clock::now();
+  start_yaw_ = tf2::getYaw(initial_pose_.orientation);
 
   return Status::SUCCEEDED;
 }
 
 Status Spin::onCycleUpdate()
 {
-  // Currently only an open-loop time-based controller is implemented
-  // The closed-loop version 'controlledSpin()' has not been fully tested
-  return timedSpin();
-}
+  geometry_msgs::msg::Pose current_pose;
+  if (!getRobotPose(current_pose)) {
+    RCLCPP_ERROR(node_->get_logger(), "Current robot pose is not available.");
+    return Status::FAILED;
+  }
+
+  const double current_yaw = tf2::getYaw(current_pose.orientation);
+  const double yaw_diff = start_yaw_ - current_yaw;
 
-Status Spin::timedSpin()
-{
-  // Output control command
   geometry_msgs::msg::Twist cmd_vel;
+  cmd_vel.linear.x = 0.0;
+  cmd_vel.linear.y = 0.0;
+  cmd_vel.angular.z = 0.0;
 
-  // TODO(orduno) #423 fixed time
-  auto current_time = std::chrono::system_clock::now();
-  if (current_time - start_time_ >= 6s) {  // almost 180 degrees
+  if (abs(yaw_diff) >= abs(command_yaw_)) {
```

Comment:
Is the yaw range pi, -pi rad? If so, the discontinuity would cause some issues here. For example if robot is commanded to move +1 rad and is currently at pi rad, then yaw would jump to some negative value on the next cycle update, let's say -pi rad. Then yaw_diff would be 2pi rad and stop the robot (when it hasn't moved the full +1 rad). Similarly, if start is -pi rad and directed to move -1 rad, the next loop update could yield yaw_diff = -2pi rad.